contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+# Resource Classes Proposal
+
+  1. [Abstract](#abstract)
+  2. [Motivation](#motivation)
+  3. [Use Cases](#use-cases)
+  4. [Objectives](#objectives)
+  5. [Non Objectives](#non-objectives)
+  6. [Resource Class](#resource-class)
+  7. [API Changes](#api-changes)
+  8. [Scheduler Changes](#sch-changes)
+  9. [Kubelet Changes](#kubelet-changes)
+ 10. [Opaque Integer Resources](#oir)
+ 11. [Future Scope](#future-scope)
+
+_Authors:_
+
+* @vikaschoudhary16 - Vikas Choudhary &lt;vichoudh@redhat.com&gt;
+* @aveshagarwal - Avesh Agarwal &lt;avagarwa@redhat.com&gt;
+
+## Abstract
+In this document we will describe *resource classes* which is a new model to
+represent compute resources in Kubernetes. This document should be seen as a
+successor to [device plugin proposal](https://github.com/kubernetes/community/pull/695/files)
+and has a dependency on the same.
+
+## Motivation
+Kubernetes system knows only two resource types, 'CPU' and 'Memory'. Any other
+resource can be requested by pod using opaque-integer-resource(OIR) mechanism.
+OIR is a key-value pair with the key being a string and the value being a 'Quantity'
+which can (optionally) be fractional. The current model is great for supporting
+simple compute resources like CPU or Memory, which are available across all
+kubernetes deployments.
+But there is a problem in representing resources like GPUs, ASICs, local storage
+etc in the form of OIRs. Each of such resource type generally has a rich
+metadata like version, capabilities etc. to describe the resource. A particular
+application pod may perform well only with resources which have a certain
+capability, for example GPUs greater than version 'V'. OIR does not allow such,
+metadata based, selection/filtering of resources.
+The current model requires identity mapping between available resources and requested
+resources. Identity mapping means there must be a one-to-one mapping between the
+resource reference used in the spec to request the resource and resource reference
+used to advertise the resource availablity.
+
+Since 'CPU' and 'Memory' are resources that are available across all kubernetes
+deployments and need no metadata to describe the resource type any further,
```

Comment:
We've seen use case where people want kubernetes to schedule pods based on different types of CPUs, e.g. run critical application on newer/better version of CPU. Is this something within the scope of resource class?  I read comment below, it seems resource claas only cares about resources advertised by device plugin.  We could of course write one, but i'm not quite sure how it interacts with the standard integer based CPU scheduling.

Reply:
> Is this something within the scope of resource class? 

Yes, but not in the first phase. Plan is to go incrementally and first handling only plugin advertised resources.